### PR TITLE
xray: add plist

### PR DESCRIPTION
**Pull request by xiruizhao**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Review by sinoban — DISMISSED** [review]

**Comment by kidonng** [comment]
Please provide a default config file like [v2ray](https://github.com/Homebrew/homebrew-core/blob/d4d579e9a140923250c481e8fdd82fddd70a942e/Formula/v2ray.rb#L37).

**Comment by carlocab** [comment]
It would also be good to bump the revision here.

File: Formula/xray.rb [hunk 1]
```diff
@@ -41,6 +46,44 @@ def install
...
+          <key>RunAtLoad</key>
+          <true/>
+          <key>Label</key>
+          <string>#{plist_name}</string>
+          <key>ProgramArguments</key>
+          <array>
+            <string>#{opt_bin}/xray</string>
+            <string>run</string>
+            <string>--config</string>
+            <string>#{etc}/xray/config.json</string>
```
Comment (SMillerDev) [inline Formula/xray.rb:82]: Does it not run without specifying this config?

File: Formula/xray.rb [hunk 2]
```diff
@@ -41,6 +46,44 @@ def install
...
+    resource("example_config").stage do
+      pkgetc.install "vless_tcp_xtls.json" => "example_config.json"
+    end
+  end
+
+  def caveats
+    <<~EOS
+      An example_config.json is installed in #{etc}/xray.
+      For more, visit https://github.com/XTLS/Xray-examples.
+      `brew services` requires a config file at #{etc}/xray/config.json
```
Comment (SMillerDev) [inline Formula/xray.rb:59]: That means the install won't work out-of-the-box. I'd rather have the install create a config.json by default if it needs any (and none is present)

File: Formula/xray.rb (hunk 1 above)
Comment (xiruizhao) [reply Formula/xray.rb:82]: It will otherwise try to read config from stdin

Comment (carlocab) [reply Formula/xray.rb:82]: So `brew services start xray` will wait for input from stdin? That's not good, and will definitely leave many users of this formula quite confused. A default config that will work out of the box is needed here, I think.

File: Formula/xray.rb (hunk 2 above)
Comment (xiruizhao) [reply Formula/xray.rb:59]: To use its functionality (network proxy), manual configuration is required. The example_config.json is for demo purpose (not necessary for installation) and will not work.

File: Formula/xray.rb (hunk 1 above)
Comment (xiruizhao) [reply Formula/xray.rb:82]: By specifying plist this way, a fresh installation will cause `brew services start xray` to error, saying that `${etc}/xray/config.json` doesn’t exist.

File: Formula/xray.rb (hunk 2 above)
Comment (kidonng) [reply Formula/xray.rb:59]: That's not the point. Installing the example config as `pkgetc/”config.json”` is better than telling the user to, say, copy the example file by themselves.

Comment (xiruizhao) [reply Formula/xray.rb:59]: But if the example file is installed as config.json and the user modified it, an update would overwrite the modification? How should this be fixed?

Comment (kidonng) [reply Formula/xray.rb:59]: It won't overwrite the config file if it's modified.

Comment (xiruizhao) [reply Formula/xray.rb]: Okay I didn't know that. Would an update modify geoip? So homebrew can distinguish user-modified file versus an outdated file, right? I made the changes.

File: Formula/xray.rb [hunk 3]
```diff
@@ -25,6 +25,11 @@ class Xray < Formula
     sha256 "b85355097f660164a65723223aaef0a62c671266d4a18927c5c12c32c036cc54"
   end
 
+  resource "example_config" do
+    url "https://raw.githubusercontent.com/XTLS/Xray-examples/main/VLESS-TCP-XTLS-WHATEVER/config_client/vless_tcp_xtls.json"
```
Comment (SMillerDev) [inline Formula/xray.rb:29]: This should come from a tagged release. As soon as the file is changed the build will fail.

File: Formula/xray.rb [hunk 4]
```diff
@@ -41,6 +47,43 @@ def install
...
+
+    resource("example_config").stage do
+      pkgetc.install "config.json"
+    end
+  end
+
+  def caveats
+    <<~EOS
+      Please modify the example config installed at #{etc}/xray/config.json
+      For some examples, see https://github.com/XTLS/Xray-examples.
```
Comment (SMillerDev) [inline Formula/xray.rb:59]:
```suggestion
      An example config is installed to #{etc}/xray/config.json
```

**Review by chenrui333 — DISMISSED** [review]
lgtm

**Comment by carlocab** [comment]
Needed rebasing to publish the bottle.

**Review by carlocab — CHANGES_REQUESTED** [review]
Can you use the new syntax for `plist`s? Example:

https://github.com/Homebrew/homebrew-core/blob/7df2ebbab3abada5937665cec631277ea3f75ae8/Formula/php.rb#L323-L330

Also, please bump the `revision` (either add `revision 1` or increment the `revision` right after `license`) so that users get the updated version of the formula when they do `brew upgrade`. Squash your changes with your first commit when you do this.

Thanks!

**Comment by carlocab** [comment]
Seems like this needs https://github.com/Homebrew/brew/pull/11355.

File: Formula/xray.rb [hunk 5]
```diff
@@ -41,6 +48,23 @@ def install
...
+    end
+  end
+
+  def caveats
+    <<~EOS
+      An example config is installed to #{etc}/xray/config.json
+    EOS
+  end
+
+  plist_options manual: "xray run --config #{etc}/xray/config.json"
```
Comment (Misaka13514) [inline Formula/xray.rb:63]: remove this line
`plist_options manual:` isn't needed

**Review by Misaka13514 — CHANGES_REQUESTED** [review]

**Comment by xiruizhao** [comment]
> Seems like this needs [Homebrew/brew#11355](https://github.com/Homebrew/brew/pull/11355), and that has to land in a release tag first.

@carlocab It's merged, could you re-review this PR?

File: Formula/xray.rb [hunk 6]
```diff
@@ -41,6 +48,22 @@ def install
     resource("geosite").stage do
       pkgshare.install "dlc.dat" => "geosite.dat"
     end
+
+    resource("example_config").stage do
+      pkgetc.install "config.json"
+    end
```
Comment (carlocab) [inline Formula/xray.rb:54]:
```suggestion
    pkgetc.install resource("example_config")
```
I think this should do the same thing; can you check this locally? We may want to simplify the lines that install other resources too if it does.

Other than this, this LGTM

**Review by carlocab — COMMENTED** [review]
Thanks for your patience here. Just one question. Feel free to ping me again when you're ready.

File: Formula/xray.rb (hunk 6 above)
Comment (xiruizhao) [reply Formula/xray.rb]: passed local tests

File: Formula/xray.rb [hunk 7]
```diff
@@ -34,13 +41,23 @@ def install
...
       XRAY_LOCATION_ASSET: "${XRAY_LOCATION_ASSET:-#{pkgshare}}"
 
-    resource("geoip").stage do
-      pkgshare.install "geoip.dat"
-    end
-
+    pkgshare.install resource("geoip")
     resource("geosite").stage do
       pkgshare.install "dlc.dat" => "geosite.dat"
     end
```
Comment (carlocab) [inline Formula/xray.rb:47]:
```suggestion
    pkgshare.install resource("geosite") => "geosite.dat"
```
Does this not work?

Comment (xiruizhao) [reply Formula/xray.rb:47]: Didn't know that. Let me check.

Comment (xiruizhao) [reply Formula/xray.rb:47]:
```
Error: An exception occurred within a child process:
  TypeError: no implicit conversion of Resource into String
```

Comment (carlocab) [reply Formula/xray.rb:47]: Ok, that's fine. We can keep this one.

**Review by carlocab — APPROVED** [review]

**Comment by carlocab** [comment]
Thanks, @xiruizhao. GitHub Actions is suffering a partial outage at the moment, so we'll probably need to re-run CI later when it's back up.

**Pull request closed without merging**